syntax = "proto3";

package surveillance;

message CanonicalTrade {
  int64 event_time_ns = 1;
  string venue = 2;
  string instrument = 3;
  string side = 4;
  double price = 5;
  double quantity = 6;
  string order_id = 7;
  string execution_id = 8;
  string participant_id = 9;
  string origin = 10; // CEX | DEX
}

message TradeBatch {
  repeated CanonicalTrade trades = 1;
}

message Ack {
  bool success = 1;
  string message = 2;
}

message Empty {}

service TradeStream {
  rpc PublishTrades (TradeBatch) returns (Ack);
  rpc Subscribe (Empty) returns (stream CanonicalTrade);
}
